doc/pins/pins_hld.md

```diff
@@ -0,0 +1,299 @@
+
+### Familiar Interface
+
+The family of P4 programs models the SAI forwarding pipeline, making the P4RT interface familiar to anyone who has worked with SONiC/SAI before. The P4RT interface provides  essential networking features (L2 bridging, L3 routing, ACLs, etc.) allowing users to quickly get up to speed on how to use this new API.
+
+### Unambiguous Documentation
+
+The SAI P4 programs will serve as an unambiguous documentation of the SAI pipeline. This will help to minimize differences between SAI implementation provided by various vendors, and serve as a reference for the community at large. The P4 program for L3 routing will be modeled after the SAI pipeline for L3. More details on the full SAI pipeline are available in the [SAI Pipeline Object Model][sai-model-ref].
+
+## Requirements
+
+The following components are targeted for the SONiC 2021-11 Release:
+
+* P4RT application that runs in its own container,
+* P4RT-Orchagent that runs as part of SWSS, and
+* SAI P4 program.
+
+These components will enable the following functionality:
+
+* P4RT clients can use P4RT to program IP route entries, next hop members and groups, and ACL entries to drop or punt packets to the control plane.
+* P4RT clients can program ACLs to punt packets received in the ingress pipeline. These packets will be punted to the P4RT application running in the switch and will be forwarded to the client over the P4RT gRPC channel.
+* Users can introduce custom private extensions to the pipeline by adding other elements to the P4 program.
+
+Additional functionality will be made available in future releases. Warmboot / fastboot capability will not be supported for the P4RT application in the first MVP release. Support will be added in subsequent releases. Warmboot / fastboot capabilities for other applications will continue to function as normal and will not be impacted by PINS.
+
+For the first release, the P4RT application will be integrated using SONiC’s traditional, monolithic build, but we will explore moving the new application extension framework in a subsequent release.
+
+## Architecture
+
+SONiC is structured into various containers that communicate through multiple logical databases via a shared Redis instance. To add SDN support, PINS introduces a few new components into the SONiC system:
+
+* P4RT: An application that receives P4 programming requests from the controller and programs the requests to the APPL DB.
+* P4RT tables: a new APPL DB table that defines the P4 programming requests.
+* P4Orch: A new orch that programs the P4RT table from APPL DB to ASIC DB. It also sends response notifications to P4RT and manages the APPL STATE DB.
+* APPL STATE DB: A new DB for applications to query the system status. It has the same schema of the APPL DB.
```

Comment:
Please change the name of this new DB? RESPONSE DB? or any other better name.

Reply:
Will change to "APPL Response DB". Having the APPL keyword as part of the DB is to indicate that the schema of the DB reflects the APPL DB and hence it is easy for an application to compare the entry present in this DB with the intent to get an understanding of which parts of the intent have been satisfied. An intent might be partially met and that would be reflected in this DB.